src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -82,153 +90,198 @@ public static bool IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLock
+                                                // Do not add private assets for processing.
+                                                if (visitedP2PReference.Add(reference.ProjectUniqueName) && reference.PrivateAssets != LibraryIncludeFlags.All)
+                                                {
+                                                    var referenceSpec = dgSpec.GetProjectSpec(reference.ProjectUniqueName);
+                                                    queue.Enqueue(new Tuple<string, string>(referenceSpec.Name, reference.ProjectUniqueName));
+                                                }
                                             }
                                         }
+                                        else // This should never happen.
+                                        {
+                                            throw new Exception(string.Format(CultureInfo.CurrentCulture, Strings.PackagesLockFile_RestoreMetadataMissingTfms));
+                                        }
                                     }
-                                    else // This should never happen.
+                                    else
                                     {
-                                        return false;
+                                        invalidReasons.Add(string.Format(
+                                               CultureInfo.CurrentCulture,
+                                               Strings.PackagesLockFile_ProjectReferenceHasNoCompatibleTargetFramework,
+                                               p2pProjectName,
+                                               framework.FrameworkName.GetShortFolderName()
+                                               ));
                                     }
                                 }
-                                else
+                                else // This can't happen. When adding the queue, the referenceSpec HAS to be discovered. If the project is otherwise missing, it will be discovered in HasP2PDependencyChanged
                                 {
-                                    return false;
+                                    throw new Exception(string.Format(
+                                        CultureInfo.CurrentCulture,
+                                        Strings.PackagesLockFile_UnableToLoadPackagespec,
+                                        p2pUniqueName));
                                 }
                             }
-                            else
-                            {
-                                return false;
-                            }
                         }
                     }
                 }
             }
 
-            return true;
+            // Validate the runtimes for the current project did not change.
+            var projectRuntimesKeys = project.RuntimeGraph.Runtimes.Select(r => r.Key).Where(k => k != null);
```

Comment:
I'd move this back up top before the individual framework checks. Reduce the amount of changes.

Reply:
moved the logic above right before we start checks for every target framework.